test_conformance/basic/test_work_item_functions.cpp

```diff
@@ -35,146 +39,552 @@ struct work_item_data
-"    uint localID[ 3 ];\n"
-"    uint numGroups[ 3 ];\n"
-"    uint groupID[ 3 ];\n"
-" } work_item_data;\n"
-"\n"
-"__kernel void sample_kernel( __global work_item_data *outData )\n"
-"{\n"
-"    int id = get_global_id(0);\n"
-"   outData[ id ].workDim = (uint)get_work_dim();\n"
-"    for( uint i = 0; i < get_work_dim(); i++ )\n"
-"   {\n"
-"       outData[ id ].globalSize[ i ] = (uint)get_global_size( i );\n"
-"       outData[ id ].globalID[ i ] = (uint)get_global_id( i );\n"
-"       outData[ id ].localSize[ i ] = (uint)get_local_size( i );\n"
-"       outData[ id ].localID[ i ] = (uint)get_local_id( i );\n"
-"       outData[ id ].numGroups[ i ] = (uint)get_num_groups( i );\n"
-"       outData[ id ].groupID[ i ] = (uint)get_group_id( i );\n"
-"   }\n"
-"}";
+const char *workItemKernelCode =
+    R"(typedef struct {
+    uint workDim;
+    uint globalSize[ 3 ];
+    uint globalID[ 3 ];
+    uint localSize[ 3 ];
+    uint localID[ 3 ];
+    uint numGroups[ 3 ];
+    uint groupID[ 3 ];
+    uint globalOffset[ 3 ];
+    uint enqueuedLocalSize[ 3 ];
+ } work_item_data;
 
-#define NUM_TESTS 1
+__kernel void sample_kernel( __global work_item_data *outData )
+{
+    int id = get_global_id(0);
+   outData[ id ].workDim = (uint)get_work_dim();
+    for( uint i = 0; i < get_work_dim(); i++ )
+   {
+       outData[ id ].globalSize[ i ] = (uint)get_global_size( i );
+       outData[ id ].globalID[ i ] = (uint)get_global_id( i );
+       outData[ id ].localSize[ i ] = (uint)get_local_size( i );
+       outData[ id ].localID[ i ] = (uint)get_local_id( i );
+       outData[ id ].numGroups[ i ] = (uint)get_num_groups( i );
+       outData[ id ].groupID[ i ] = (uint)get_group_id( i );
+   }
+})";
+
+const char *outOfRangeWorkItemKernelCode =
+    R"(typedef struct {
+    uint workDim;
+    uint globalSize[ 3 ];
+    uint globalID[ 3 ];
+    uint localSize[ 3 ];
+    uint localID[ 3 ];
+    uint numGroups[ 3 ];
+    uint groupID[ 3 ];
+    uint globalOffset[ 3 ];
+    uint enqueuedLocalSize[ 3 ];
+ } work_item_data;
+
+__kernel void sample_kernel( __global work_item_data *outData )
+{
+    int ind_mul=1;
+    int ind=0;
+    for( uint i = 0; i < get_work_dim(); i++ )
+    {
+        ind += (uint)get_global_id(i) * ind_mul;
+        ind_mul *= get_global_size(i);
+    }
+    outData[ind].workDim = (uint)get_work_dim();
+%s
+})";
+
+const char *outOfRangeWorkItemKernelCodeExt =
+    R"(
+    uint dimindx=get_work_dim()+1;
+    outData[ind].globalSize[0] = (uint)get_global_size(dimindx);
+    outData[ind].globalID[0] = (uint)get_global_id(dimindx);
+    outData[ind].localSize[0] = (uint)get_local_size(dimindx);
+    outData[ind].localID[0] = (uint)get_local_id(dimindx);
+    outData[ind].numGroups[0] = (uint)get_num_groups(dimindx);
+    outData[ind].groupID[0] = (uint)get_group_id(dimindx);
+)";
 
-int test_work_item_functions(cl_device_id deviceID, cl_context context, cl_command_queue queue, int num_elements)
+const char *outOfRangeWorkItemKernelCodeExt11 =
```

Comment:
Could we use a single kernel for this stuff, and guard the parts that are specific to newer OpenCL C versions using `__OPENCL_VERSION__` or some other preprocessor mechanism?  I think that would be easier to understand and maintain vs. stitching together these sample kernels with `sprintf`.